samples/GalleryApp/MainPage.xaml.cs

```diff
@@ -21,29 +16,80 @@ namespace GalleryApp
     /// </summary>
     public sealed partial class MainPage : Page
     {
+        public ObservableCollection<Skill> SkillPages { get; } = new ObservableCollection<Skill>();
+
+        private static List<SkillCategory> allCategories;
+        private static SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
         public MainPage()
         {
             this.InitializeComponent();
         }
 
         /// <summary>
-        /// Navigation to Skeletal Detector Page
+        /// Invoked when the Page is loaded and becomes the current source of a parent Frame.
         /// </summary>
-        /// <param name="sender"></param>
         /// <param name="e"></param>
-        private void NavigateToSkeletalDetectorPage(object sender, RoutedEventArgs e)
+        protected override void OnNavigatedTo(NavigationEventArgs e)
         {
-            this.Frame.Navigate(typeof(SkeletalDetectorPage));
+            if (SkillPages.Count == 0)
+            {
+                LoadAllSkills();
+            }
+            base.OnNavigatedTo(e);
         }
 
         /// <summary>
-        /// Navigation to Object Detector Page
+        /// Load all the skills available
         /// </summary>
-        /// <param name="sender"></param>
-        /// <param name="e"></param>
-        private void NavigateToObjectDetectorPage(object sender, RoutedEventArgs e)
+        private async void LoadAllSkills()
+        {
+            // Task: bind category to UI tab
+            var SkillsCategory = await GetCategoriesAsync();
+            foreach (var category in SkillsCategory)
+            {
+                foreach (var skill in category.Skills)
+                {
+                    SkillPages.Add(skill);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Navigate to a specific skill page
+        /// </summary>
+        private void SampleGridView_ItemClick(object sender, ItemClickEventArgs e)
+        {
+            NavigateToSkillPage(e.ClickedItem as Skill);
+        }
+
+        /// <summary>
+        /// Navigate to the specific skill page
+        /// </summary>
+        /// <param name="skill"> Skill passed from the thumbnail item </param>
+        public void NavigateToSkillPage(Skill skill)
         {
-            this.Frame.Navigate(typeof(ObjectDetectorPage));
+            this.Frame.Navigate(skill.PageType);
+        }
+
+        public static async Task<List<SkillCategory>> GetCategoriesAsync()
+        {
+            await _semaphore.WaitAsync();
```

Comment:
>await _semaphore.WaitAsync(); [](start = 12, length = 29)

@louis @mariana I referenced the Microsoft Community Toolkit's implementation of this function, and am not sure if using thread is necessary because loading the file is the only one task here?